ConfigurationModel deserializer and management command

cahrens opened the pull request:
### Description
 
[TNL-4781](https://openedx.atlassian.net/browse/TNL-4781)

Also https://openedx.atlassian.net/browse/TNL-4782

Adds general deserialization and management command for ConfigurationModels.

### Testing
- [x] i18n
- [x] safecommit shows 0 violations
- [x] Unit, integration, acceptance tests as appropriate

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [ ] Code review: @cpennington 
- [ ] Code review: @efischer19 

FYI: @ssemenova, @feanil and/or @maxrothman -- I will be using this management command to create Ansible scripting for setting xblock support state on prod/stage/sandboxes.

### Post-review
- [ ] Squash commits

cahrens commented on common/djangoapps/config_models/utils.py:
```diff
@@ -0,0 +1,46 @@
+        stream_or_string = stream_or_string.decode('utf-8')
+    try:
+        objects = json.loads(stream_or_string)
+        change_date = unicode(timezone.now())
```
Lines 27-28 are the only modifications to this code from https://github.com/django/django/blob/7f51876f99851fdc3fef63aecdfbcffa199c26b9/django/core/serializers/json.py#L70

cpennington replied:
If you explicitly `save`d the object, then it would add the `change_date`, right?

cahrens replied:
No, it doesn't. If I don't specify change_date at all it complains, and if I specify anything that isn't a real date string, it also complains.

cpennington replied:
Strange. I wonder why https://docs.djangoproject.com/en/1.9/ref/models/fields/#django.db.models.DateField.auto_now_add doesn't work.

cahrens replied:
If I end up going with this approach, I'll dig into it more.

cahrens commented on common/djangoapps/config_models/utils.py:
```diff
@@ -0,0 +1,46 @@
+    model_name = objects["model"]
+    data = objects["data"]
+    for fields in data:
+        fields[u"change_date"] = change_date
```
change_date has to be specified (and I want it specified to know when the management command was run)

Unfortunately I don't see anyway to get useful changed_by information. The None will make it obvious that it was updated via management command, and we will have to use the github history of the config file to see who authorized a particular change.

cpennington replied:
In the modulestore, I believe we have a couple of known bogus user ids that are used for this type of thing. Another option would be to require it to be logged as a service user.

cahrens replied:
How does a management command require you to be logged in?

cpennington replied:
Logged, not logged in. We'd need to either hard-code a user id, allow it to be passed in in the command, or figure out a way to identify the services users in the database and query for the right one. I'd probably lean towards the user_id argument option, at the moment.

cahrens commented on common/djangoapps/config_models/test/test_utils.py:
```diff
@@ -0,0 +1,63 @@
+    def setUp(self):
+        super(ConfigurationModelTests, self).setUp()
+
+    def test_deserialization(self):
```
Minimal test to show how this works.

cpennington commented on common/djangoapps/config_models/test/test_utils.py:
```diff
@@ -0,0 +1,63 @@
+    def test_deserialization(self):
+        self.assertEquals(0, len(ExampleDeserializeConfig.objects.all()))
+        start_date = timezone.now()
+        test_json = ('{'
```
You could use `textwrap.dedent(""" <some multi-line-string> """)` so that you don't need to have quotes on every line.

cpennington commented:
We have a DjangoRestFramework serializer that we use for the API to manage ConfigModels. Would that be any cleaner?

cahrens commented:
@cpennington you mean this code?

````
class AutoConfigModelSerializer(ModelSerializer):
    """Serializer class for configuration models."""
    class Meta(object):
        """Meta information for AutoConfigModelSerializer."""
        model = self.model
````
It felt wrong to use a serializer from the Django rest framework, but I'll give it a try.

cpennington commented:
Yeah, that's the one. I feel like using the same format that's used by the API is probably less confusing, rather than more.

cahrens commented:
OK, I'll look into using AutoConfigModelSerializer and let you know how that goes.

cahrens commented:
@cpennington I've switched to using the Django REST framework serializer. It correctly handles the change_date, which is nice. I've also added the ability to specify a user id.

Please take a look and let me know what you think (I still need to add additional test cases before official code review). Thanks!

cpennington commented:
Looks good to me.

cahrens commented on common/djangoapps/config_models/migrations/0001_add_config_models_service_user.py:
```diff
@@ -0,0 +1,25 @@
+from django.db import migrations
+
+
+USERNAME = settings.CONFIG_MODEL_SERVICE_WORKER_USERNAME
```
This is the same thing done [here](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/commerce/migrations/0001_data__add_ecommerce_service_user.py). Please advise if there is a better approach.

cpennington commented on common/djangoapps/config_models/management/commands/populate_model.py:
```diff
@@ -0,0 +1,72 @@
+        if not os.path.exists(json_file):
+            raise CommandError(_("File {0} does not exist").format(json_file))
+
+        print _("Importing JSON data from file {0}").format(json_file)
```
Per the django docs, it's better to use `self.stdout.write` than `print` in management commands (see https://docs.djangoproject.com/en/1.8/howto/custom-management-commands/)

cahrens commented on common/djangoapps/config_models/tests/tests.py:
```diff
@@ -104,6 +109,53 @@ def test_always_insert(self, __):
+
+    def test_equality(self):
+        config = ExampleConfig(changed_by=self.user, string_field='first')
+        config.cache_timeout = 0
```
@cpennington Existing unit tests used a mock_cache, which I did not want because it caused my new methods to get mocked out.

cpennington replied:
Why are your new methods mocked out when using a mock cache? Assuming you make the mock cache always return `None`, it shouldn't affect your code.

cahrens commented on common/djangoapps/config_models/utils.py:
```diff
@@ -0,0 +1,69 @@
+    if list_serializer.is_valid():
+        model_class = serializer_class.Meta.model
+        for data in reversed(list_serializer.validated_data):
+            if model_class.equal_to_existing_entry(data):
```
@cpennington Note that this was a bit more complicated than we discussed because the serializer creates and saves model instances in a single operation, in the "create" method, which is called by "save". Thus I had to create model instances for comparison (but not save them). See https://github.com/tomchristie/django-rest-framework/blob/master/rest_framework/serializers.py#L872

cpennington commented on common/djangoapps/config_models/models.py:
```diff
@@ -176,3 +176,46 @@ def key_values(cls, *key_fields, **kwargs):
+        Compares this instance to a model instance constructed from the supplied JSON for equality.
+        This will ignore any fields in `fields_to_ignore`.
+
+        Note that this method cannot handle fields with many-to-many associations, as those can only
```
Should this method automatically skip those Many-To-Many associations?

cahrens replied:
I could have it do that. I couldn't think of a use case for a ConfigurationModel with many-to-many associations.

cpennington replied:
Yeah, that's fair. I'm not sure I can either.

cahrens replied:
I went ahead and wrote the code (and a unit test) to ignore the many-to-many relationships. I'm not entirely sure that is the desired behavior (perhaps it would be better to error if there are many-to-many relationships?), but it can always be revisited if it turns out that such a use case does exist.

cpennington commented on common/djangoapps/config_models/models.py:
```diff
@@ -176,3 +176,46 @@ def key_values(cls, *key_fields, **kwargs):
+        return True
+
+    @classmethod
+    def equal_to_existing_entry(cls, json, fields_to_ignore=("id", "change_date", "changed_by")):
```
Nit: perhaps "equal_to_current"?

cahrens commented:
@cpennington and @efischer19 If you have time to review this PR tomorrow, I would very much appreciate it-- thank you!

efischer19 commented on common/djangoapps/config_models/management/commands/populate_model.py:
```diff
@@ -0,0 +1,72 @@
+
+        self.stdout.write(_("Importing JSON data from file {0}").format(json_file))
+        with open(json_file) as data:
+            created_entries = deserialize_json(data, options['username'])
```
Should this be wrapped in a try, in case of improperly formatted json?

efischer19 commented on common/djangoapps/config_models/utils.py:
```diff
@@ -0,0 +1,69 @@
+        list_serializer.save()
+        return entries_created
+    else:
+        raise Exception(list_serializer.error_messages)
```
Nvm, got ahead of myself with that last comment. Got it now.

efischer19 commented on common/djangoapps/config_models/models.py:
```diff
@@ -176,3 +179,58 @@ def key_values(cls, *key_fields, **kwargs):
+
+        Note that this method cannot handle fields with many-to-many associations, as those can only
+        be set on a saved model instance (and saving the model instance will create a new entry).
+        All many-to-many field entries will be removed before for the equality comparison is done.
```
"before for"

efischer19 commented:
👍 1 minor docstring note, looks great otherwise! Very thorough with the test coverage.

cpennington commented on common/djangoapps/config_models/management/commands/populate_model.py:
```diff
@@ -0,0 +1,72 @@
+
+    JSON should be in a file, with the following format:
+
+    { "model": "config_models.exampleconfigurationmodel",
```
Should the class name be TitleCased?

cahrens replied:
This was the format used by the Django serializer. I can try title case to see if it also works.

I also wondered why it isn't config_models_exampleconfigurationmodel, but again, I was just starting with something that I knew worked.

cpennington replied:
If this is a standard django format, then that's fine.

cahrens replied:
This is what the doc string for the django.apps.registry.get_model method says:

````
Returns the model matching the given app_label and model_name.

        As a shortcut, this function also accepts a single argument in the
        form <app_label>.<model_name>.

        model_name is case-insensitive.
````

Since model_name is documented to be case-insensitive, I'm fine TitleCasing it to match the declaration of the model.

cpennington commented on common/djangoapps/config_models/utils.py:
```diff
@@ -0,0 +1,69 @@
+    list_serializer = serializer_class(data=parsed_json["data"], context={"changed_by_username": username}, many=True)
+    if list_serializer.is_valid():
+        model_class = serializer_class.Meta.model
+        for data in reversed(list_serializer.validated_data):
```
Why is this reversed? Is that so that `.remove` doesn't break the iteration? Would something like:

    list_serializer.validated_data = [data for data in list_serializer.validated_data if model_class.equal_to_current(data)]

be clearer?

cahrens replied:
Yes, it was to prevent breaking the iteration. If I recall, part of the issue was that I can't assign to list_serializer.validated_data. I'll double-check about this though.

cpennington replied:
Ok.

cahrens replied:
Yes, validated_data is a property without a setter.

````
AttributeError: can't set attribute

cpennington commented on common/djangoapps/config_models/models.py:
```diff
@@ -176,3 +179,58 @@ def key_values(cls, *key_fields, **kwargs):
+            if relation_info.to_many and (field_name in json):
+                json.pop(field_name)
+
+        new_instance = cls(**json)
```
It seems a bit odd to create a model instance here. Could `equal_to_current` just compare the fields on `cls.current(*key_field_args)` with the items in `json`? Or does the class constructor do data conversion on them?

cahrens replied:
The class constructor does conversion, plus I would then have to account for default values of fields (if no value is specified in the JSON, which is a very common situation). I thought it was much easier to just create the instance to ensure I'm comparing apples to apples, so to speak.

cpennington replied:
Makes sense to me.

cpennington commented:
I had a few small comments, but am :+1: once you resolve them one way or another.

cahrens commented:
jenkins run bokchoy
